configs/config_row_changes_few.py-3000.txt
# Mutating Neo
doing 1000000 prerandomization switches.
doing 3000 random mutations
worse ['eh'] - 0 / 3000
worse ['eo'] - 1 / 3000
worse ['xt'] - 2 / 3000
worse ['fb'] - 3 / 3000
worse ['fg'] - 4 / 3000
worse ['ip'] - 5 / 3000
worse ['uv'] - 6 / 3000
9431.28466502 ['kß'] finger repetition: 25.115061 position cost: 1746.074126
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'u', 'ü', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', '.', 'l', 'c', 'r', 'y', 'g', 'p', 't', 'n', 'k', '⇘', '\n'], ['⇧', (), 'b', 'ß', 'm', ',', 'o', 'ä', 'ö', 'i', 'w', 'h', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 7 / 3000
worse ['kr'] - 8 / 3000
worse ['mk'] - 9 / 3000
worse ['fw'] - 10 / 3000
worse ['hn'] - 11 / 3000
worse ['hp'] - 12 / 3000
worse ['lg'] - 13 / 3000
worse ['kn'] - 14 / 3000
worse ['cg'] - 15 / 3000
worse ['eß'] - 16 / 3000
worse ['zo'] - 17 / 3000
9208.98560642 ['yu'] finger repetition: 24.681135 position cost: 1698.270196
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ü', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', '.', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'k', '⇘', '\n'], ['⇧', (), 'b', 'ß', 'm', ',', 'o', 'ä', 'ö', 'i', 'w', 'h', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 18 / 3000
9097.33480444 ['bi'] finger repetition: 20.175828 position cost: 1715.03429
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ü', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', '.', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'k', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'o', 'ä', 'ö', 'b', 'w', 'h', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 19 / 3000
worse ['xj'] - 20 / 3000
7499.25309189 ['kh'] finger repetition: 20.175828 position cost: 1674.985845
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ü', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', '.', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'o', 'ä', 'ö', 'b', 'w', 'k', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 21 / 3000
7403.34533407 ['äü'] finger repetition: 20.011996 position cost: 1675.200307
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ä', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', '.', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'o', 'ü', 'ö', 'b', 'w', 'k', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 22 / 3000
worse ['ür'] - 23 / 3000
worse ['ne'] - 24 / 3000
worse ['bi'] - 25 / 3000
worse ['gc'] - 26 / 3000
worse ['ps'] - 27 / 3000
worse ['tm'] - 28 / 3000
worse ['gk'] - 29 / 3000
worse ['tö'] - 30 / 3000
worse ['rä'] - 31 / 3000
7390.79589707 ['b.'] finger repetition: 16.759836 position cost: 1664.352603
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ä', 'x', 'v', 'a', 'j', 'd', 'e', 'f', 'z', 's', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'o', 'ü', 'ö', '.', 'w', 'k', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 32 / 3000
7323.0681191 ['of'] finger repetition: 14.802387 position cost: 1631.167473
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ä', 'x', 'v', 'a', 'j', 'd', 'e', 'o', 'z', 's', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'k', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 33 / 3000
worse ['il'] - 34 / 3000
worse ['uä'] - 35 / 3000
worse ['lö'] - 36 / 3000
worse ['wg'] - 37 / 3000
worse ['gz'] - 38 / 3000
worse [',ü'] - 39 / 3000
worse ['vü'] - 40 / 3000
worse ['jx'] - 41 / 3000
worse ['ss'] - 42 / 3000
worse ['dj'] - 43 / 3000
7278.60795185 ['kv'] finger repetition: 15.87279 position cost: 1626.102653
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ä', 'x', 'k', 'a', 'j', 'd', 'e', 'o', 'z', 's', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 'g', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 44 / 3000
7221.39824027 ['gs'] finger repetition: 17.638045 position cost: 1550.161427
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'y', 'ä', 'x', 'k', 'a', 'j', 'd', 'e', 'o', 'z', 'g', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 45 / 3000
worse ['ww'] - 46 / 3000
worse ['cg'] - 47 / 3000
7038.52238837 ['jy'] finger repetition: 17.577838 position cost: 1550.161427
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'j', 'ä', 'x', 'k', 'a', 'y', 'd', 'e', 'o', 'z', 'g', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 48 / 3000
worse ['ßf'] - 49 / 3000
worse ['wö'] - 50 / 3000
worse ['hp'] - 51 / 3000
worse ['sl'] - 52 / 3000
worse ['cä'] - 53 / 3000
worse ['ö,'] - 54 / 3000
7030.74374224 ['ky'] finger repetition: 16.518377 position cost: 1561.883163
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'j', 'ä', 'x', 'y', 'a', 'k', 'd', 'e', 'o', 'z', 'g', '´', ()], ['⇩', 'q', 'b', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 55 / 3000
worse ['go'] - 56 / 3000
worse ['nt'] - 57 / 3000
worse ['go'] - 58 / 3000
worse ['n,'] - 59 / 3000
6850.35384985 ['bq'] finger repetition: 17.163204 position cost: 1561.883163
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'j', 'ä', 'x', 'y', 'a', 'k', 'd', 'e', 'o', 'z', 'g', '´', ()], ['⇩', 'b', 'q', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 60 / 3000
worse ['ji'] - 61 / 3000
6725.08094854 ['kj'] finger repetition: 16.907589 position cost: 1561.883163
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'y', 'a', 'j', 'd', 'e', 'o', 'z', 'g', '´', ()], ['⇩', 'b', 'q', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'm', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 62 / 3000
worse ['yü'] - 63 / 3000
worse ['un'] - 64 / 3000
worse ['äs'] - 65 / 3000
worse ['dz'] - 66 / 3000
worse ['äi'] - 67 / 3000
worse ['dk'] - 68 / 3000
worse ['gö'] - 69 / 3000
worse ['oö'] - 70 / 3000
worse ['iy'] - 71 / 3000
worse ['vß'] - 72 / 3000
worse ['äx'] - 73 / 3000
worse ['uo'] - 74 / 3000
worse ['qß'] - 75 / 3000
worse ['ßn'] - 76 / 3000
worse ['sä'] - 77 / 3000
6719.64812238 ['mo'] finger repetition: 17.33196 position cost: 1566.877213
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'y', 'a', 'j', 'd', 'e', 'm', 'z', 'g', '´', ()], ['⇩', 'b', 'q', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'o', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 78 / 3000
worse ['nc'] - 79 / 3000
worse ['äx'] - 80 / 3000
worse ['sp'] - 81 / 3000
worse ['bl'] - 82 / 3000
worse ['fy'] - 83 / 3000
worse ['uz'] - 84 / 3000
6627.17605919 ['yz'] finger repetition: 17.212386 position cost: 1561.378565
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'a', 'j', 'd', 'e', 'm', 'y', 'g', '´', ()], ['⇩', 'b', 'q', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'o', ',', 'f', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 85 / 3000
worse ['ef'] - 86 / 3000
worse ['lv'] - 87 / 3000
6615.37051021 ['fg'] finger repetition: 16.068811 position cost: 1571.886143
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'a', 'j', 'd', 'e', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'q', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'o', ',', 'g', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 88 / 3000
worse ['äe'] - 89 / 3000
6548.82718624 ['oq'] finger repetition: 15.075403 position cost: 1520.714673
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'a', 'j', 'd', 'e', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'o', 'l', 'c', 'r', 'u', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'q', ',', 'g', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 90 / 3000
worse ['hz'] - 91 / 3000
6341.4820596 ['ua'] finger repetition: 16.232983 position cost: 1500.135669
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'e', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'o', 'l', 'c', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'q', ',', 'g', 'ü', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 92 / 3000
worse ['pc'] - 93 / 3000
6170.39088185 ['ül'] finger repetition: 19.835867 position cost: 1532.351557
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'e', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'o', 'ü', 'c', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'i', 'ß', 'q', ',', 'g', 'l', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 94 / 3000
worse ['üü'] - 95 / 3000
worse ['x,'] - 96 / 3000
worse ['xq'] - 97 / 3000
worse ['ho'] - 98 / 3000
worse ['äo'] - 99 / 3000
worse ['xw'] - 100 / 3000
worse ['rl'] - 101 / 3000
worse ['on'] - 102 / 3000
5909.47101527 ['io'] finger repetition: 19.157964 position cost: 1407.304855
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'e', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'i', 'ü', 'c', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'ß', 'q', ',', 'g', 'l', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 103 / 3000
worse ['ib'] - 104 / 3000
worse ['öß'] - 105 / 3000
worse ['r,'] - 106 / 3000
worse ['äx'] - 107 / 3000
worse ['ua'] - 108 / 3000
worse ['bü'] - 109 / 3000
5779.56086528 ['ie'] finger repetition: 18.850487 position cost: 1384.527128
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'c', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'ß', 'q', ',', 'g', 'l', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 110 / 3000
worse ['üb'] - 111 / 3000
worse ['am'] - 112 / 3000
5766.71653695 ['ßl'] finger repetition: 18.802056 position cost: 1431.676658
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'c', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'l', 'q', ',', 'g', 'ß', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 113 / 3000
worse ['bi'] - 114 / 3000
worse ['nu'] - 115 / 3000
worse [',ä'] - 116 / 3000
worse ['ää'] - 117 / 3000
worse ['ßs'] - 118 / 3000
worse ['ßy'] - 119 / 3000
worse ['rz'] - 120 / 3000
worse ['yr'] - 121 / 3000
worse ['yr'] - 122 / 3000
worse ['vr'] - 123 / 3000
5698.07051201 ['cl'] finger repetition: 16.131635 position cost: 1408.806974
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'u', 'j', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'l', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', 'q', ',', 'g', 'ß', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 124 / 3000
worse ['ag'] - 125 / 3000
worse ['fj'] - 126 / 3000
worse ['qe'] - 127 / 3000
worse ['öd'] - 128 / 3000
worse ['nw'] - 129 / 3000
worse ['ef'] - 130 / 3000
worse ['nö'] - 131 / 3000
worse ['fä'] - 132 / 3000
worse ['dj'] - 133 / 3000
worse ['of'] - 134 / 3000
worse [',t'] - 135 / 3000
worse ['oo'] - 136 / 3000
worse ['aü'] - 137 / 3000
worse ['yz'] - 138 / 3000
worse ['wb'] - 139 / 3000
worse ['ur'] - 140 / 3000
worse ['oo'] - 141 / 3000
worse ['ss'] - 142 / 3000
worse ['on'] - 143 / 3000
worse ['yß'] - 144 / 3000
5402.98003033 ['uj'] finger repetition: 15.732472 position cost: 1418.308821
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'j', 'u', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'l', 'r', 'a', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', 'q', ',', 'g', 'ß', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 145 / 3000
worse ['cv'] - 146 / 3000
worse ['kz'] - 147 / 3000
worse ['.s'] - 148 / 3000
worse ['ik'] - 149 / 3000
5385.59095646 ['la'] finger repetition: 15.120573 position cost: 1408.069237
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'j', 'u', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'a', 'r', 'l', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', 'q', ',', 'g', 'ß', 'ö', '.', 'w', 'v', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 150 / 3000
worse ['fz'] - 151 / 3000
worse ['ta'] - 152 / 3000
worse ['me'] - 153 / 3000
worse ['iz'] - 154 / 3000
worse ['hk'] - 155 / 3000
worse ['ee'] - 156 / 3000
worse ['xd'] - 157 / 3000
worse ['ü.'] - 158 / 3000
worse ['.z'] - 159 / 3000
worse ['vk'] - 160 / 3000
5138.25178086 ['v.'] finger repetition: 16.813033 position cost: 1406.893831
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'j', 'u', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ü', 'a', 'r', 'l', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', 'q', ',', 'g', 'ß', 'ö', 'v', 'w', '.', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 161 / 3000
worse ['xä'] - 162 / 3000
worse ['z,'] - 163 / 3000
worse ['nf'] - 164 / 3000
worse ['m,'] - 165 / 3000
worse ['wd'] - 166 / 3000
5084.80563363 ['üö'] finger repetition: 16.928943 position cost: 1410.226558
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'j', 'u', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ö', 'a', 'r', 'l', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', 'q', ',', 'g', 'ß', 'ü', 'v', 'w', '.', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 167 / 3000
worse ['öh'] - 168 / 3000
worse ['ya'] - 169 / 3000
worse ['lh'] - 170 / 3000
worse ['lj'] - 171 / 3000
worse ['i,'] - 172 / 3000
worse ['bö'] - 173 / 3000
4938.0743253 ['q,'] finger repetition: 16.568501 position cost: 1410.226558
[['^', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '-', '`', '←'], ['⇥', 'k', 'ä', 'x', 'z', 'j', 'u', 'd', 'i', 'm', 'y', 'f', '´', ()], ['⇩', 'b', 'e', 'ö', 'a', 'r', 'l', 's', 'p', 't', 'n', 'h', '⇘', '\n'], ['⇧', (), 'o', 'c', ',', 'q', 'g', 'ß', 'ü', 'v', 'w', '.', '⇗'], [(), (), (), ' ', (), (), (), ()]]
- 174 / 3000
worse ['bv'] - 175 / 3000
worse ['rh'] - 176 / 3000
worse ['aq'] - 177 / 3000
worse ['cy'] - 178 / 3000
worse ['bz'] - 179 / 3000
worse ['qh'] - 180 / 3000
worse ['or'] - 181 / 3000
worse ['üd'] - 182 / 3000
worse ['äa'] - 183 / 3000
worse ['är'] - 184 / 3000
worse ['br'] - 185 / 3000
worse ['rc'] - 186 / 3000
worse ['yu'] - 187 / 3000
worse ['hü'] - 188 / 3000
worse ['qf'] - 189 / 3000
worse ['jf'] - 190 / 3000
worse ['äd'] - 191 / 3000
worse ['tk'] - 192 / 3000
worse ['tk'] - 193 / 3000
worse ['wy'] - 194 / 3000
worse ['km'] - 195 / 3000
worse ['ud'] - 196 / 3000
worse ['df'] - 197 / 3000
worse ['kv'] - 198 / 3000
worse ['sk'] - 199 / 3000
worse ['uß'] - 200 / 3000

config.py
#!/usr/bin/env python3

"""Configuration of check_neo, mainly the weights of the cost functions. Intended to be easily modified."""

#: The mutated letters - only these get changed.
abc = "abcdefghijklmnopqrstuvwxyzäöüß,."

WEIGHT_POSITION = 1 #: reference cost - gets multiplied with the COST_PER_KEY.

WEIGHT_FINGER_REPEATS = 32 #: Cost of a finger repeat. Gets additional +1 from the row change on the same finger.

WEIGHT_FINGER_REPEATS_TOP_BOTTOM = 64 #: Additional cost of a finger repetition from the top to the bottom line. Gets added to the cost of the normal finger repetition. Additionally this gets costs as row repetition on the same hand (+4). 

WEIGHT_BIGRAM_ROW_CHANGE_PER_ROW = 1 #: When I have to switch the row in a bigram while on the same hand, that takes time => Penalty per (row to cross / horizontal distance)² if we’re on the same hand. 

WEIGHT_COUNT_ROW_CHANGES_BETWEEN_HANDS = False #: Should we count a row change with a handswitch as row change? 

WEIGHT_FINGER_DISBALANCE = 30 #: multiplied with the standard deviation of the finger usage - value guessed and only valid for the 1gramme.txt corus. 

WEIGHT_TOO_LITTLE_HANDSWITCHING = 1 #: how high should it be counted, if the hands aren’t switched in a triple?

WEIGHT_NO_HANDSWITCH_AFTER_DIRECTION_CHANGE = 5 #: how much stronger should the triple without handswitch be counted, if there also is a direction change? Also affects the “unweighted” result from total_cost!

WEIGHT_INTENDED_FINGER_LOAD_LEFT_PINKY_TO_RIGHT_PINKY = [
    1,
    2,
    2,
    2.6, # is 1/3 faster
    1,
    1,
    2.6,
    2,
    2,
    1] #: The intended load per finger. Inversed and then used as multiplier for the finger load before calculating the finger disbalance penalty. Any load distribution which strays from this optimum gives a penalty.

WEIGHT_XCVZ_ON_BAD_POSITION = 0.6 #: the penalty *per letter* in the text if xvcz are on bad positions (cumulative; if all 4 are on bad positions (not in the first 5 keys, counted from the left side horizontally) we get 4 times the penalty). 

WEIGHT_FINGER_SWITCH = 1 #: how much worse is it to switch from middle to indexfinger compared with middle to pinky (~30ms according to Rohmert).

#: The cost for moving from one finger to another one with middle-to-index as 1 (30ms). Currently only uses the neighbors. Can also be used to favor a certain dairection. Adapted the Rohmert times as per my own experiences: http://lists.neo-layout.org/pipermail/diskussion/2010-May/017171.html and http://lists.neo-layout.org/pipermail/diskussion/2010-May/017321.html
FINGER_SWITCH_COST = {
    "Klein_L": {"Ring_L": 2}, # 100ms
    "Ring_L": {"Klein_L": 3,
               "Mittel_L": 5}, # 140ms
    "Mittel_L": {"Ring_L": 4,
                 "Zeige_L": 1}, # Nach Rohmert 230ms statt 200ms ⇒ 30ms
    "Zeige_L": {"Mittel_L": 4}, # 120ms
    "Daumen_L": {},
    "Daumen_R": {},
    "Zeige_R": {"Mittel_R": 4},
    "Mittel_R": {"Zeige_R": 1,
                 "Ring_R": 4},
    "Ring_R": {"Mittel_R": 5,
               "Klein_R": 3}, 
    "Klein_R": {"Ring_R": 2}
}

WEIGHT_NO_HANDSWITCH_AFTER_UNBALANCING_KEY = 1 #: How much penalty we want if there’s no handswitching after an unbalancing key. Heavy unbalancing (wkßz, M3 right, return and the shifts) counts double (see UNBALANCING_POSITIONS). This also gives a penalty for handswitching after an upparcase letter.

#: Positions which pull the hand from the base row, position and cost (the strength of the pulling from base row). 
UNBALANCING_POSITIONS = {
(1, 0, 0): 2, # Tab
(1, 1, 0): 1,# x
(1, 5, 0): 2, # w
(1, 6, 0): 2, # k
(1, 10, 0): 1, # q
(1, 11, 0): 2, # ß
(2, 0, 0): 2, # L_M3
(2, 5, 0): 1, # o
(2, 6, 0): 1, # s
(2, 11, 0): 1, # y
(2, 12, 0): 2, # R_M3
(2, 13, 0): 2, # Return
(3, 0, 0): 2, # L_Shift
(3, 12, 0): 2, # R_Shift
(3, 6, 0): 2 # z
}

# Structured key weighting (but still mostly from experience and deducing from the work of others).
# The speed of the fingers is taken out (see INTENDED_FINGER_LOAD_LEFT_PINKY_TO_RIGHT_PINKY).
# So this shows the reachability of the keys, ignoring the different speed of the fingers.
# “How much does the hand hurt when reaching for the key” :)
# rationale: reaching for the Neo 2 x hurts thrice as much as just hitting the Neo2 u → 10 vs. 3.
# the upper keys on the right hand can be reached a bit better, since movement is aligned with the hand
# (except for q, since the pinky is too short for that).
# theoretical minimum (assigning the lowest cost to the most frequent char, counting only the chars on layer 1):
# 1123111113 = 3.3490913205386508 mean key position cost
# Ringfinger on lower row takes 1.5 times the time of index on the upper row[1].
# [1]: http://forschung.goebel-consult.de/de-ergo/rohmert/Rohmert.html - only one person!
COST_PER_KEY  = [ # the 0 values aren’t filled in at the moment. 
    [0,     0, 0, 0, 0, 0,     0, 0, 0, 0, 0,22, 0, 0], # Zahlenreihe (0)
    [0,    10, 6, 5, 6, 9,    10, 5, 4, 5, 8,12,18, 0], # Reihe 1
    [0,     3, 3, 3, 3, 5,     5, 3, 3, 3, 3, 5,10,18], # Reihe 2
    [15,10,12,12,10, 10,   15, 7, 6,11,11,10,   15],     # Reihe 3
    [0,0,0,               3           , 0, 0, 0, 0] # Reihe 4 mit Leertaste
]

# for reference the neo layout
NEO = [
    [("^"),("1"),("2"),("3"),("4"),("5"),("6"),("7"),("8"),("9"),("0"),("-"),("`"),("←")], # Zahlenreihe (0)
    [("⇥"),("x"),("v"),("l"),("c"),("w"),("k"),("h"),("g"),("f"),("q"),("ß"),("´"),()], # Reihe 1
    [("⇩"),("u"),("i"),("a"),("e"),("o"),("s"),("n"),("r"),("t"),("d"),("y"),("⇘"),("\n")], # Reihe 2
    [("⇧"),(),("ü"),("ö"),("ä"),("p"),("z"),("b"),("m"),(","),("."),("j"),("⇗")],        # Reihe 3
    [(), (), (), (" "), (), (), (), ()] # Reihe 4 mit Leertaste
]


#: The positions which are by default accessed by the given finger. 
FINGER_POSITIONS = {
    "Klein_L": [(1, 1, 0), (2, 0, 0), (2, 1, 0), (3, 0, 0), (3, 1, 0), (3, 2, 0)], # Klein_L
    "Ring_L": [(1, 2, 0), (2, 2, 0), (3, 3, 0)], # Ring_L
    "Mittel_L": [(1, 3, 0), (2, 3, 0), (3, 4, 0)], # Mittel_L
    "Zeige_L": [(1, 4, 0), (2, 4, 0), (3, 5, 0), (1, 5, 0), (2, 5, 0), (3, 6, 0)], # Zeige_L
    "Daumen_L": [(4, 3, 0)], # Daumen_L
    "Daumen_R": [(4, 3, 0)], # Daumen_R
    "Zeige_R": [(1, 6, 0), (2, 6, 0), (3, 7, 0), (1, 7, 0), (2, 7, 0), (3, 8, 0)], # Zeige_R
    "Mittel_R": [(1, 8, 0), (2, 8, 0), (3, 9, 0)], # Mittel_R
    "Ring_R": [(1, 9, 0), (2, 9, 0), (3, 10, 0)], # Ring_R
    "Klein_R": [(1, 10, 0), (2, 10, 0), (3, 11, 0), (1, 11, 0), (2, 11, 0), (1, 12, 0), (2, 12, 0), (2, 13, 0), (3, 12, 0)] # Klein_R
}
